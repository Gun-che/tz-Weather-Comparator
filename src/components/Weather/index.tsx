import React from 'react'
import s from './index.module.scss'

export const Weather: React.FC = () => {
  return (
    <section className={s.wrap}>
      Weather
    </section>
  )
}

export default Weather