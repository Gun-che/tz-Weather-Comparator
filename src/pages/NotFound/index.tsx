import React from 'react'
import s from './index.module.scss'

export const NotFound: React.FC = () => {
  return (
    <>
      <div className={s.wrap + ' containerCust'}>
        404
        <h3>not found</h3>
      </div>
    </>
  )
}


export default NotFound