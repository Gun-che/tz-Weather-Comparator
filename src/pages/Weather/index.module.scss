@import "../../SASS/const";
.df,
.cw,
.hf {
  transition: all 0.4s;
  color: $dark;
  &:hover,
  &:focus {
    background-color: rgba($blue, $alpha: 0.3);
    a {
      color: $dark - #444 !important;
    }
  }
}

.df .df {
  background: $blue;
  a {
    color: white !important;
  }
}
.cw .cw {
  background: $blue;
  a {
    color: white !important;
  }
}
.hf .hf {
  background: $blue;
  a {
    color: white !important;
  }
}

.tabs {
  display: flex;
  margin: 20px auto;
  width: 100%;
  justify-content: center;
  &:hover {
    background-color: transparent;
  }
  ul {
    width: auto;
    margin: 0 auto;
    align-items: center;
  }
}

.form {
  display: flex;
  > div {
    width: 100%;
  }
  button {
    margin: auto 0 auto 20px;
    line-height: 0;
    padding: 0;
    width: 44px;
    height: 40px;
    margin: 15px 0 15px 20px;
  }
}

@media all and (max-width: 600px) {
  .tabs {
    &:hover {
      background-color: transparent;
    }
    ul {
      flex-direction: column;
      height: auto;
    }
  }

  .form {
    > div {
      width: 80%;
    }
    button {
      margin: 15px 0 15px auto;
    }
  }
}

@media all and (max-width: 400px) {
  .form {
    > div {
      width: 70%;
    }
    button {
      margin: 15px 0 15px auto;
    }
  }
}
