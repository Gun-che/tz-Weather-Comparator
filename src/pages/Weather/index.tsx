import React, { SyntheticEvent } from 'react'
import { Route, Link, useLocation } from 'react-router-dom'
import { PropsFromRedux } from '../../containers/WeatherContainer'
import { CurrentWeather } from './components/CurrentWeather'
import { errorMsg } from '../../helpers/messages'
import { HourlyForecast } from './components/HourlyForecast'
import { LoadingThin } from '../../components/Loading'
import { DailyForecast } from './components/DailyForecast'

import s from './index.module.scss'
import { ErrorMessage } from '../../components/ErrorMessage'

export const Weather: React.FC<PropsFromRedux> = ({
  getWeather,
  daily,
  hourly,
  data,
  message,
  code,
  isFetching
}) => {

  const tmpLocation = useLocation()
  console.log(tmpLocation)

  const [searchValue, setSearchValue] = React.useState<string>('')
  const [formMessage, setFormMessage] = React.useState<string>('')
  const [tabsClass, setTabsClass] = React.useState(s.cw)

  React.useEffect(() => {
    const loc = tmpLocation.pathname;
    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
    loc === '/weather' && tabsClass !== s.cw ? setTabsClass(s.cw) :
      loc === '/weather/hourly' ? setTabsClass(s.hf) :
        loc === '/weather/daily' ? setTabsClass(s.df) :
          null

  }, [tabsClass, tmpLocation.pathname])

  const onSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (searchValue) {
      getWeather({ city: searchValue })
      formMessage && setFormMessage('')

    } else {
      setFormMessage(errorMsg.empty['ru'])
    }
  }

  const onChange = (e: SyntheticEvent<HTMLInputElement>) => {
    setSearchValue(e.currentTarget.value)
  }

  const tmpMsg = (): string => {
    const fullMsg = errorMsg[code] ? errorMsg[code]['ru'] : message
    return fullMsg || formMessage ? `${fullMsg} ${formMessage}` : ''
  }

  return (
    <section className={s.wrap + ' container'}>
      <ErrorMessage message={tmpMsg()} />
      <div className={tabsClass + ' card-tabs ' + s.tabs}>
        <ul className="tabs">
          <li className={s.cw + " tab"}>
            <Link to="/weather">Текущая погода</Link>
          </li>
          <li className={s.hf + " tab"}>
            <Link to="/weather/hourly">Почасовой прогноз</Link>
          </li>
          <li className={s.df + " tab"}>
            <Link to="/weather/daily">Прогноз на неделю</Link>
          </li>
        </ul>
      </div>
      <form className={s.form} onSubmit={onSubmit}>
        <div className="input-field">
          <input
            name="search"
            type="text"
            id="search"
            value={searchValue}
            onChange={onChange}
            className="validate"
          />
          <label htmlFor="search">Введите город</label></div>
        <button
          type="submit"
          className="waves-effect waves-light btn-floating blue lighten-3">
          <i className="material-icons left">search</i></button>
      </form>

      <Route path="/weather" exact>
        {data.length ?
          <CurrentWeather data={data[0]} /> :
          isFetching ? <LoadingThin /> :
            <h2 className={s.input}>Введите название города</h2>}
      </Route>

      <Route path="/weather/hourly">
        {hourly.length ?
          <HourlyForecast data={hourly} name={data[0].name} /> :
          isFetching ? <LoadingThin /> :
            <h2 className={s.input}>Введите название города</h2>}
      </Route>

      <Route path="/weather/daily">
        {daily.length ?
          <DailyForecast name={data[0].name} data={daily} /> :
          isFetching ? <LoadingThin /> :
            <h2 className={s.input}>Введите название города</h2>}
      </Route>
    </section>
  )
}