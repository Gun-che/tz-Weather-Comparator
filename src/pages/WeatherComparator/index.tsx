import React, { SyntheticEvent, ChangeEvent } from 'react'

import { PropsFromRedux } from '../../containers/WeatherComparatorContainer'
import { errorMsg } from '../../helpers/messages'
import { ComparisonCard } from '../../components/ComparisonCard'
import { WeatherCard } from '../../components/WeatherCard'
import { CityInput } from '../../components/CityInput'
import { ComparisonInfo } from '../../components/ComparisonInfo'
import s from './index.module.scss'
import { ErrorMessage } from '../../components/ErrorMessage'

export const WeatherComparator: React.FC<PropsFromRedux> = ({
  data,
  getWeather,
  message,
  code,
}) => {

  const [cities, setSities] = React.useState<[string, number][]>([['', 0], ['', 1]])
  const [formMessage, setFormMessage] = React.useState<string>('')
  const [key, setKey] = React.useState<number>(2)

  const onSubmit = (e: SyntheticEvent<HTMLFormElement>) => {
    e.preventDefault();

    if (cities.some(i => i[0])) {
      const post = cities
        .filter(i => i[0])
        .map(i => i[0])

      getWeather(post)
      formMessage && setFormMessage('')

    } else {
      setFormMessage(errorMsg.empty2['ru'])
    }
  }

  const onChange = (e: ChangeEvent<HTMLInputElement & { dataset: { index: number } }>) => {
    const target = e.currentTarget;
    let arr = [...cities];
    arr[target.dataset.index][0] = target.value;

    setSities(arr);
  }

  const onAdd = (e: SyntheticEvent) => {
    setSities([...cities, ['', key]])
    setKey(key + 1)
  }

  const onClose = (i: number): (e: SyntheticEvent) => void => e => {
    let arr = [...cities];
    arr.splice(i, 1);
    setSities(arr)
  }


  const tmpGridClass = (): string => {
    return (data.length === 2 || data.length === 4) ? s.two :
      data.length === 0 ? s.empty : ''
  }

  const tmpMsg = (): string => {
    const fullMsg = errorMsg[code] ? errorMsg[code]['ru'] : message
    return fullMsg || formMessage ?
      `${fullMsg} ${formMessage}` :
      ''
  }

  return (
    <div className={s.wrap + ' ' + tmpGridClass() + ' containerCust'}>

      <div className={s.message}>
        <h2>Введите названия городов, погоду в которых вы хотите сравнить</h2>
        <ErrorMessage message={tmpMsg()} />
      </div>

      <form onSubmit={onSubmit} className={s.form}>
        {cities.map((item, index) => {
          return <div key={item[1]}>
            <CityInput
              onClose={onClose}
              item={item}
              index={index}
              onChange={onChange}
            />
          </div>
        })}
        <div className={s.btns}>
          <button
            onClick={onAdd}
            className="waves-effect waves-light btn blue lighten-3">
            <i className="material-icons left">add_location</i>Add location</button>
          <button
            type="submit"
            className="waves-effect waves-light btn blue lighten-3">
            <i className="material-icons left">search</i>Search</button>
        </div>
      </form>

      {data.length ?
        <><ComparisonCard data={data} />
          <div className={s.infoBlock}><ComparisonInfo data={data} /></div>
        </> : null}

      <div className={s.grid}>
        {data.map((item, index) => {
          return (
            <div key={index}>
              <WeatherCard data={item} />
            </div>
          )
        })}
      </div>
    </div>
  )
}

