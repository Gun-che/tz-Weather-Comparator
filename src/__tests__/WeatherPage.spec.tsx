import React from 'react'
import { mount } from 'enzyme'
import { BrowserRouter as R } from 'react-router-dom'
import { Weather } from '../pages/Weather'

describe('weatherPage component', () => {
  const props = {
    getWeather: jest.fn(),
    daily: [],
    hourly: [],
    data: [],
    message: '',
    code: 1,
    isFetching: false,
  }
  describe('weatherPage component render', () => {
    const weatherPage = mount(<R><Weather {...props} /></R>)

    it('renders properly', () => {
      expect(weatherPage).toMatchSnapshot()
    })

    it('render <Route>', () => {
      expect(weatherPage.find('Route')).toHaveLength(3)
    })

    it('render <Link>', () => {
      expect(weatherPage.find('Link')).toHaveLength(3)
    })

    it('render <form>', () => {
      expect(weatherPage.find('form')).toHaveLength(1)
    })
  })
})