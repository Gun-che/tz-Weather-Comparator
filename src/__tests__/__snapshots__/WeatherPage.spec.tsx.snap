// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`weatherPage component weatherPage component render renders properly 1`] = `ReactWrapper {}`;
